src/collection-view.js

```diff
@@ -101,12 +101,21 @@ Marionette.CollectionView = Marionette.View.extend({
   _onCollectionAdd: function(child, collection, opts) {
     // `index` is present when adding with `at` since BB 1.2; indexOf fallback for < 1.2
     var index = opts.at !== undefined && (opts.index || collection.indexOf(child));
+    var models = collection.models;
 
     // When filtered or when there is no initial index, calculate index.
-    if (this.getOption('filter') || index === false) {
-      index = _.indexOf(this._filteredSortedModels(index), child);
+    if (this.getOption('filter') || this.getViewComparator() || index === false) {
+      models = this._filteredSortedModels(index);
     }
 
+    // find adjusted index
+    _.reduce(models, function(counter, model) {
+      var view = this.children.findByModel(model);
+      if (view) { return view._index; }
+
+      return model === child ? index = counter + 1 : counter;
+    }, -1, this);
```

Comment:
@JSteunou the model index is not identical to the view `_index`. Imagine this use case quickly: I have a collection with two models:

```
{ val: 3 } # model index: 0 # view _index: 0
{ val: 4 } # model index: 1 # view _index: 1
```

now I add two more models:

```
{ val: 2 } # model index: 1
{ val: 1 } # model index: 0
```

At the point where the patch was introduced the state of the system looks like:

```
# we are inside of onCollectionAdd
# child = model where { val: 2 }

{ val: 1 } # model index: 0
{ val: 2 } # model index: 1
{ val: 3 } # model index: 2 # view _index: 0
{ val: 4 } # model index: 3 # view _index: 1
```

Now if we utilized only the model index as our "index" it would insert `{ val: 2 }` in the DOM in between `{ val: 3 }` and `{ val: 4 }`. We must utilize the view index to ensure the node is added at the correct location in the DOM....